src/app/app-platform/ContentAppPlatform.cpp

```diff
@@ -100,37 +100,67 @@ EndpointId ContentAppPlatform::AddContentApp(ContentApp * app, EmberAfEndpointTy
         if (mContentApps[index] == app)
         {
             ChipLogProgress(DeviceLayer, "Already added");
+            // already added, return endpointId of already added endpoint.
+            // desired endpointId does not have any impact
             return app->GetEndpointId();
         }
         index++;
     }
 
+    if (desiredEndpointId >= emberAfFixedEndpointCount() &&
+        emberAfGetDynamicIndexFromEndpoint(desiredEndpointId) != kEmberInvalidEndpointIndex)
+    {
+        // an endpoint already exists with the desiredEndpointId. return error.
+        ChipLogError(DeviceLayer, "Failed to add dynamic endpoint: desired endpointID already in use!");
+        return kNoCurrentEndpointId;
+    }
+
     index = 0;
     while (index < CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT)
     {
         if (nullptr == mContentApps[index])
         {
             mContentApps[index] = app;
             EmberAfStatus ret;
-            while (1)
+
+            if (desiredEndpointId >= emberAfFixedEndpointCount())
             {
-                ret = emberAfSetDynamicEndpoint(index, mCurrentEndpointId, ep, dataVersionStorage, deviceTypeList);
+                ret = emberAfSetDynamicEndpoint(index, desiredEndpointId, ep, dataVersionStorage, deviceTypeList);
                 if (ret == EMBER_ZCL_STATUS_SUCCESS)
                 {
                     ChipLogProgress(DeviceLayer, "Added ContentApp %s to dynamic endpoint %d (index=%d)", vendorApp.applicationId,
-                                    mCurrentEndpointId, index);
-                    app->SetEndpointId(mCurrentEndpointId);
+                                    desiredEndpointId, index);
+                    app->SetEndpointId(desiredEndpointId);
                     return app->GetEndpointId();
                 }
-                else if (ret != EMBER_ZCL_STATUS_DUPLICATE_EXISTS)
+                else
                 {
                     ChipLogError(DeviceLayer, "Adding ContentApp error=%d", ret);
                     return kNoCurrentEndpointId;
                 }
-                // Handle wrap condition
-                if (++mCurrentEndpointId < mFirstDynamicEndpointId)
+            }
+            else
+            {
+                while (1)
                 {
-                    mCurrentEndpointId = mFirstDynamicEndpointId;
+                    ret = emberAfSetDynamicEndpoint(index, mCurrentEndpointId, ep, dataVersionStorage, deviceTypeList);
+                    if (ret == EMBER_ZCL_STATUS_SUCCESS)
+                    {
+                        ChipLogProgress(DeviceLayer, "Added ContentApp %s to dynamic endpoint %d (index=%d)",
+                                        vendorApp.applicationId, mCurrentEndpointId, index);
+                        app->SetEndpointId(mCurrentEndpointId);
```

Comment:
once we use mCurrentEndpointId, would it make sense to increment it? Or do we rely that it will get incremented later with wrap handling on line 160?

Reply:
makes sense to increment. it would be handled by the duplicate endpoint handling but it is better to be explicit.